airbyte-cdk/python/airbyte_cdk/sources/declarative/schema/first_record_schema_loader.py

```diff
@@ -0,0 +1,97 @@
+#
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+#
+
+from dataclasses import InitVar, dataclass
+from typing import Any, Mapping, Union, Dict, List
+import dateutil.parser
+
+from airbyte_cdk.sources.declarative.schema.schema_loader import SchemaLoader
+from airbyte_cdk.sources.declarative.retrievers.retriever import Retriever
+
+
+def is_valid_date_or_datetime_or_time(string: str) -> bool:
+    try:
+        dateutil.parser.parse(string)
+        return True
+    except ValueError:
+        return False
+
+
+def is_datetime(string: str) -> bool:
+    try:
+        dt = dateutil.parser.parse(string)
+        dt.time()
+        dt.date()
+        return True
+    except ValueError:
+        return False
+
+
+def is_time(string: str) -> bool:
+    try:
+        dt = dateutil.parser.parse(string)
+        dt.time()
+        return True
+    except ValueError:
+        return False
+
+
+def is_date(string: str) -> bool:
+    try:
+        dt = dateutil.parser.parse(string)
+        dt.date()
+        return True
+    except ValueError:
+        return False
+
+
+def map_value_to_schema_dtype(value: Any) -> Dict[str, Union[Dict[str, Any], str, List[Any]]]:
+    if isinstance(value, bool):
+        return {"type": ["boolean", "null"]}
+    elif isinstance(value, int) or isinstance(value, float):
+        return {"type": ["number", "null"]}
+    elif isinstance(value, list):
+        return {"type": ["array", "null"]}
+    elif isinstance(value, str) and is_valid_date_or_datetime_or_time(value):
+        if is_datetime(value):
+            return {"type": ["string", "null"], "format": "date-time"}
+        if is_date(value):
+            return {"type": ["string", "null"], "format": "date"}
+        if is_time(value):
+            return {"type": ["string", "null"], "format": "time"}
+        return {"type": ["string", "null"]}
+    elif isinstance(value, str):
+        return {"type": ["string", "null"]}
+    elif isinstance(value, dict):
+        nested_schema: Dict[str, Any] = {"properties": {}, "type": ["object", "null"]}
+        for k, v in value.items():
+            nested_schema["properties"][k] = map_value_to_schema_dtype(v)
+        return nested_schema
+    else:
+        raise TypeError(f"Unsupported value type: {type(value)}")
+
+
+def convert_record_to_schema(record: Dict[str, Any]) -> Dict[str, Any]:
```

Comment:
I recommend leveraging the existing [SchemaInferrer class](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/utils/schema_inferrer.py#L68). It's the same code used by the connector builder